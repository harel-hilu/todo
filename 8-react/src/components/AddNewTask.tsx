import React, { ChangeEvent, useState } from 'react';
import {createUseStyles} from 'react-jss'

export default function AddNewTask(props: any) {
    const [text, setText] = useState<string>('');
    const classes: Record<string, string> = useStyles();

    function addTaskIfNotEmpty(): void {
        if (text != '') {
            props.addTask(text);
            setText('');
        }
    }

    return (
        <div className={classes.container}>
            <input className={classes.taskInput}
                autoFocus 
                value={text} 
                onChange={(e: ChangeEvent<HTMLInputElement>) => setText(e.target.value)}
                onKeyDown={(e: React.KeyboardEvent<HTMLDivElement>) => 
                    (e.key === "Enter" && addTaskIfNotEmpty())} 
            />
            <button className={classes.addButton} onClick={addTaskIfNotEmpty}>
                Add Task
            </button>
        </div>
    );  
}

const useStyles = createUseStyles({
    container: {
        display: "flex",
        marginBottom: "20px",
    },
    taskInput: {
        flexGrow: 1,
        maxWidth: "400px",
        borderRadius: "5px",
        paddingLeft: "5px",
    },
    addButton: {
        color: "white",
        alignSelf: "flex-end",
        marginRight: "5px",
        marginLeft: "auto",
        border: "none",
        backgroundColor: "black",
        borderRadius: "5px",
        height: 24,
    },
})