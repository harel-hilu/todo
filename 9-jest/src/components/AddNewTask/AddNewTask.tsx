import React, { useState } from 'react';

export default function AddNewTask() {
    const [text, setText] = useState('');
    
    return (
        <>
            <input 
                value={text}
                onChange={e => setText(e.target.value)}
                autoFocus
            />
            <button
                onClick={() => setText('')}
            >
                Add Task!
            </button>
        </>
    );
}