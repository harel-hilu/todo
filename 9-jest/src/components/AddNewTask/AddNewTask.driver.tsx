import React from 'react';
import AddNewTask from "./AddNewTask";
import Enzyme, { mount, ReactWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { act } from 'react-dom/test-utils';
import { AddNewTaskProps } from '../../types/props';

Enzyme.configure({ adapter: new Adapter() })

export const addNewTaskDriver = () => {
    let wrapper: Enzyme.ReactWrapper<AddNewTaskProps>;
    const addTask = jest.fn(() => {});
    
    return {
        given: {
            setParentWrapper: (inputWrapper: Enzyme.ReactWrapper) => 
                wrapper = inputWrapper.find(AddNewTask),
            createAddNewTaskWrapper: () =>
                wrapper = mount(<AddNewTask addTask={addTask} />),
            updateWrapper: () => wrapper.update()
        },
        when: {
            simulateInputChange: async (text = "") => await act(async () => {
                await wrapper.find("#addTaskInput")
                    .simulate('change', { target: { value: text } });
            }),
            simulateAddTaskClick: async () => {
                await act(async () => {
                    await wrapper.find('button[children="Add Task!"]').simulate("click");
                });
                
                await wrapper.update();
            },
            simulateEnterPress: () =>
                wrapper.find("#addTaskInput").simulate("keypress", {key: "Enter"})
        },
        then: {
            addTask: addTask,
            getAddTaskButtonText: () => 
                wrapper.find('button[children="Add Task!"]').text(),
            getAddTaskInputText: () => 
                wrapper.find("#addTaskInput").prop("value")
        }
    }
}
