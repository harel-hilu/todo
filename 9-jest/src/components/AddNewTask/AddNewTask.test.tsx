import React from 'react';
import AddNewTask from "./AddNewTask";
import Enzyme, { shallow, render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders correctly enzyme', () => {
    const wrapper = shallow(<AddNewTask />)
  
    expect(toJson(wrapper)).toMatchSnapshot();
});