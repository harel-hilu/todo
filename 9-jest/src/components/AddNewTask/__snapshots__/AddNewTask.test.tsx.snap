// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly enzyme 1`] = `
<Fragment>
  <input
    autoFocus={true}
    onChange={[Function]}
    value=""
  />
  <button
    onClick={[Function]}
  >
    Add Task!
  </button>
</Fragment>
`;
